src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+  tmp_symbol.mode=ID_java;
+  tmp_symbol.name=name;
+  tmp_symbol.type=type;
+  symbol_table.add(tmp_symbol);
+  return symbol_exprt(name, type);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::declare_function
+
+  Inputs: a name and a type
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void string_refine_preprocesst::declare_function(
+  irep_idt function_name, const typet &type)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
+  func_symbol.is_static_lifetime=false;
+  func_symbol.mode=ID_java;
+  func_symbol.name=function_name;
+  func_symbol.type=type;
+  symbol_table.add(func_symbol);
+  goto_functions.function_map[function_name];
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::get_data_and_length_type_of_string
+
+  Inputs: an expression, a reference to a data type and a reference to a
+          length type
+
+ Purpose: assuming the expression is a java string, figure out what
+          the types for length and data are and put them into the references
+          given as argument
+
+\*******************************************************************/
+
+void string_refine_preprocesst::get_data_and_length_type_of_string(
+  const exprt &expr, typet &data_type, typet &length_type)
+{
+  assert(refined_string_typet::is_java_string_type(expr.type()) ||
+         refined_string_typet::is_java_string_builder_type(expr.type()));
+  typet object_type=ns.follow(expr.type());
+  assert(ns.follow(object_type).id() == ID_struct);
+  const struct_typet &struct_type=to_struct_type(object_type);
+  for(unsigned i=0; i<struct_type.components().size(); i++)
+    if(struct_type.components()[i].get_name() == "length")
+      length_type=struct_type.components()[i].type();
+    else if(struct_type.components()[i].get_name() == "data")
+      data_type=struct_type.components()[i].type();
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_cprover_string_assign
+
+  Inputs: a goto_program, a position in this program, an expression and a
+          location
+
+ Outputs: an expression
+
+ Purpose: Introduce a temporary variable for cprover strings;
+          returns the cprover_string corresponding to rhs if it is a string
+          pointer and the original rhs otherwise.
+
+\*******************************************************************/
+
+exprt string_refine_preprocesst::make_cprover_string_assign(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const exprt &rhs,
+  const source_locationt &location)
+{
+  if(refined_string_typet::is_java_string_pointer_type(rhs.type()))
+  {
+    auto it=java_to_cprover_strings.find(rhs);
+    if(it!=java_to_cprover_strings.end())
+      return it->second;
+    else
+    {
```

Comment:
As a general pattern: use insert and pick up the returned <iterator,bool> pair. If that second element is true, run through the body of your `else`. At the end of that body, update the element pointed to by the iterator. Finally just return the element pointed to by the iterator, irrespective of whether inside that `else` or not. This avoids the duplicate lookup for find and the later `[...]` operator.